### ISSUE TITLE
How to use a quaternion

### ISSUE THREAD

**Issue by sekaiwodamase** [CMM-22/a2#5]
So for many years, each time a student asks any prof about quaternions, the answer ends up being something among the lines of 'this is a complicated subject outside the scope of the course, so we won't get into it'. Now we seemingly need to use a heading quaternion in Exercise 3, but I have no idea what to do with it. Any explanation would be much appreciated!

**Comment by eastskykang** [comment]
We will revisit more details about quaternion in the lecture very soon. 

**TL;DR:**

In short, quaternion is a one way to represent rotational transformation. It's real part is ```cos (0.5 * theta)``` and it's imaginary part is ```sin(0.5 * theta) * v ``` where (theta, v) is an angle of rotation and axis of rotation respectively: https://en.wikipedia.org/wiki/Quaternion

For instance, let's say you rotate some vector along x axis by 1 rad. The corresponding quaternion to this rotational operation is ```[cos(0.5), sin(0.5), 0, 0, 0]```

------
For now, in our code base, you can just use quaternion as rotation matrix.

For instance, let's say Eigen quaternion ```qAB``` is equivalent to a rotational matrix R_{A}{B}:``v_A = R_AB v_B``` where v_A is a 3D vector **v** expressed in frame A, and v_B is **v** expressed in frame B .

Then, in cpp code, you can use quaternion as follows

```cpp
V3D v_in_frameA = qAB * v_in_frameB
```

I hope this helps.

But if you need a material to learn more about quaternion. I recommend this amazing video from 3brown1blue: https://www.youtube.com/watch?v=zjMuIxRvygQ&t=9s

**Comment by eastskykang** [comment]
By the way, you can use ```getRotationQuaternion``` function as follows to convert heading angle into quaternion. 

```cpp
Quaternion heading = getRotationQuaternion(headingAngle, V3D(0, 1, 0));
```

**Comment by sekaiwodamase** [comment]
Thank you so much Dongho for the comprehensive explanation and all the references, it is much appreciated :D
It's all clear now!
